### PR TITLE
Make bpf2c maps parser match verifier's maps parser

### LINKED ISSUE #900: bpf2c: make maps parser match verifier's maps parser
### Describe the feature you'd like supported

As noted in https://github.com/microsoft/ebpf-for-windows/pull/883

In https://github.com/vbpf/ebpf-verifier/pull/295 I put a more intelligent algorithm for determining the map definition size in the file in https://github.com/vbpf/ebpf-verifier/blob/master/src/asm_files.cpp#L79 which is the algorithm already used by libbpf.

### Proposed solution

bpf2c should change to use that algorithm, which supports older (smaller) structures by 0-padding the rest of the fields, for example by having a common implementation rather than reimplementing the parsing logic.

### Additional context

_No response_

### LINKED ISSUE #2580: Workflow failed - verifier_fuzzer
[Failed Run](https://github.com/microsoft/ebpf-for-windows/actions/runs/5231923732)
[Codebase](https://github.com/microsoft/ebpf-for-windows/tree/f734b5e1d9b3a616647015992a6804bcfdd4f123)
Test name - `verifier_fuzzer`

### PR DESCRIPTION
**Pull request by dthaler**
## Description

The verifier will parse "maps/*" sections, and is able to deal with maps sections with various record sizes.  This PR updates bpf2c to use the same algorithm.  In the future it would be good to refactor the verifier so the same code can be used.

Before this change, such a program would pass the verifier but bpf2c couldn't convert it.
Also upgrading to the latest verifier will fail without this PR.

Also the PR to add BTF support regressed the map_in_map_v2.c sample which stopped testing the legacy reference-by-id paths.  This PR changes map_in_map_v2.c back and renames it to map_in_map_legacy_id.c for clarity.  Similarly, it restores map_in_map_legacy_idx.c with the original map_in_map.c and renames the BTF map_in_map.c to map_in_map_btf.c.  This ensures all three code paths for finding the inner map are tested.

Fixes #900
Fixes #2580

Note to code reviewers: this change is easiest to review by ignoring whitespace changes (since it removes an indentation level for a bunch of code).  To do so in github, click "Files changed", then the gear button dropdown button just under the "Checks" and "Files changed" labels, click "Hide whitespace" then "Apply and reload".

## Testing

The update to map_in_map.c results in testing this functionality.

## Documentation

No impact.

### CODE REVIEW

File: tools/bpf2c/bpf_code_generator.cpp [hunk 1]
```diff
@@ -354,62 +354,116 @@ bpf_code_generator::extract_program(const bpf_code_generator::unsafe_string& sec
...
+            map_count++;
+    }
+    if (map_count == 0) {
+        return;
+    }
+
+    size_t data_size = map_section->get_size();
+    size_t map_record_size = data_size / map_count;
+
+    if (data_size % map_record_size != 0) {
```
Comment (saxena-anurag) [inline tools/bpf2c/bpf_code_generator.cpp:416]: Should we have a check before this line that `map_record_size != 0` ?

**Review by saxena-anurag — COMMENTED** [review]
I think there are some more places in ebpf-for-windows code, where it either assumes the maps section will always be `"maps"` or hard codes the maps section size. Should we fix those also in this PR?

1. `_get_program_and_map_names()` (verifier.cpp) only looks for "maps" section to find the map names, and fails if any map name is not found in "maps" section.
2. `parse_maps_section_windows()` assumes the map size to be `sizeof(ebpf_map_definition_in_file_t)`
3. `read_elf()` (elfhelper.cpp) only ignores "maps" section, it should also ignore any "maps/*" sections.

File: tools/bpf2c/bpf_code_generator.cpp (hunk 1 above)
Comment (dthaler) [reply tools/bpf2c/bpf_code_generator.cpp:528]: Done

**Comment by dthaler** [comment]
> I think there are some more places in ebpf-for-windows code, where it either assumes the maps section will always be `"maps"` or hard codes the maps section size. Should we fix those also in this PR?
> 
> 1. `_get_program_and_map_names()` (verifier.cpp) only looks for "maps" section to find the map names, and fails if any map name is not found in "maps" section.
> 2. `parse_maps_section_windows()` assumes the map size to be `sizeof(ebpf_map_definition_in_file_t)`

Thanks.  Both of those are now updated.

> 3. `read_elf()` (elfhelper.cpp) only ignores "maps" section, it should also ignore any "maps/*" sections.

ElfHelper appears to be dead code.  Filed PR #2572 for removal.

**Comment by dthaler** [comment]
Test requires https://github.com/vbpf/ebpf-verifier/pull/510

**Comment by dthaler** [comment]
blocked on https://github.com/vbpf/ebpf-verifier/issues/506

**Comment by dthaler** [comment]
sanitize pass is failing because libbtf forces /WX and forces non /Zi.
The lack of /Zi means warning C5072 is generated during the sanitize pass:
```
    28>D:\a\ebpf-for-windows\ebpf-for-windows\external\ebpf-verifier\src\crab_utils\stats.cpp : warning C5072: ASAN enabled without debug information emission. Enable debug info for better ASAN error reporting [D:\a\ebpf-for-windows\ebpf-for-windows\external\ebpf-verifier\build\ebpfverifier.vcxproj]
```
And forcing /WX means that forces the github job to fail.
Other libraries either allow /WX or /Zi to be under the control of our project.
Waiting for Alan to fix in libbtf.

**Review by Alan-Jowett — APPROVED** [review]

**Review by shpalani — COMMENTED** [review]
The curious question behind the changes:
Is the goal to support more than one verifier (from different vendors)?
Or,
Is the goal to support different versions of the same vendor's verifier?

Reason: I see the map query definition by **id**. So, how is the id determined?

**Comment by dthaler** [comment]
> The curious question behind the changes: Is the goal to support more than one verifier (from different vendors)? Or, Is the goal to support different versions of the same vendor's verifier?

No.  The goal is just to support PREVAIL and have ebpf-for-windows and prevail match in terms of what map capabilities they support.

> Reason: I see the map query definition by **id**. So, how is the id determined?

**Review by saxena-anurag — APPROVED** [review]

**Review by Alan-Jowett — APPROVED** [review]

**Pull request merged**